File: kotlin/cbor/src/main/kotlin/com/hopae/eudi/wallet/cbor/CborEncoder.kt
package com.hopae.eudi.wallet.cbor

import java.io.ByteArrayOutputStream

/**
 * Map key ordering profile.
 *
 * [Bytewise8949] is the RFC 8949 §4.2.1 core deterministic order (default).
 * [LengthFirst7049] is the legacy RFC 7049 §3.9 canonical order (length, then bytewise) —
 * kept selectable because ISO 18013-5 artifacts in the wild differ; interop to be pinned in M4.
 */
enum class MapKeyOrder { Bytewise8949, LengthFirst7049 }

object CborEncoder {

    fun encode(value: Cbor, mapOrder: MapKeyOrder = MapKeyOrder.Bytewise8949): ByteArray {
        val out = ByteArrayOutputStream()
        writeValue(value, out, mapOrder)
        return out.toByteArray()
    }

    private fun head(out: ByteArrayOutputStream, major: Int, arg: ULong) {
        val mt = major shl 5
        when {
            arg < 24u -> out.write(mt or arg.toInt())
            arg <= 0xFFu -> {
                out.write(mt or 24); out.write(arg.toInt())
            }
            arg <= 0xFFFFu -> {
                out.write(mt or 25); out.write(((arg shr 8) and 0xFFu).toInt()); out.write((arg and 0xFFu).toInt())
            }
            arg <= 0xFFFF_FFFFu -> {
                out.write(mt or 26)
                for (s in 24 downTo 0 step 8) out.write(((arg shr s) and 0xFFu).toInt())
            }
            else -> {
                out.write(mt or 27)
                for (s in 56 downTo 0 step 8) out.write(((arg shr s) and 0xFFu).toInt())
            }
        }
    }

    private fun writeValue(v: Cbor, out: ByteArrayOutputStream, order: MapKeyOrder) {
        when (v) {
            is Cbor.UInt -> head(out, 0, v.value)
            is Cbor.NInt -> head(out, 1, v.n)
            is Cbor.Bytes -> {
                head(out, 2, v.value.size.toULong()); out.write(v.value)
            }
            is Cbor.Text -> {
                val b = v.value.encodeToByteArray()
                head(out, 3, b.size.toULong()); out.write(b)
            }
            is Cbor.Array -> {
                head(out, 4, v.items.size.toULong())
                v.items.forEach { writeValue(it, out, order) }
            }
            is Cbor.CborMap -> writeMap(v, out, order)
            is Cbor.Tagged -> {
                head(out, 6, v.tag); writeValue(v.value, out, order)
            }
            is Cbor.Bool -> out.write(if (v.value) 0xF5 else 0xF4)
            Cbor.Null -> out.write(0xF6)
            Cbor.Undefined -> out.write(0xF7)
            is Cbor.Simple -> writeSimple(v.value, out)
            is Cbor.Fp -> writeFloat(v, out)
        }
    }

    private fun writeSimple(value: UByte, out: ByteArrayOutputStream) {
        val i = value.toInt()
        require(i !in 20..23) { "simple($i) has a dedicated encoding (bool/null/undefined)" }
        if (i < 20) out.write(0xE0 or i) else {
            out.write(0xF8); out.write(i)
        }
    }

    private fun writeFloat(fp: Cbor.Fp, out: ByteArrayOutputStream) {
        val d = fp.value
        if (d.isNaN()) { // canonical NaN
            out.write(0xF9); out.write(0x7E); out.write(0x00); return
        }
        val h = Half.exactBitsOf(d)
        if (h != null) {
            out.write(0xF9); out.write((h ushr 8) and 0xFF); out.write(h and 0xFF); return
        }
        val f = d.toFloat()
        if (f.toDouble().toRawBits() == fp.bits) {
            out.write(0xFA)
            val fb = f.toRawBits()
            for (s in 24 downTo 0 step 8) out.write((fb ushr s) and 0xFF)
        } else {
            out.write(0xFB)
            for (s in 56 downTo 0 step 8) out.write(((fp.bits ushr s) and 0xFF).toInt())
        }
    }

    private fun writeMap(m: Cbor.CborMap, out: ByteArrayOutputStream, order: MapKeyOrder) {
        head(out, 5, m.entries.size.toULong())
        val encoded = m.entries.map { (k, v) -> Pair(encode(k, order), v) }
        val sorted = encoded.sortedWith { a, b -> compareKeys(a.first, b.first, order) }
        for (i in 1 until sorted.size) {
            require(!sorted[i - 1].first.contentEquals(sorted[i].first)) { "duplicate map key" }
        }
        sorted.forEach { (kb, v) ->
            out.write(kb); writeValue(v, out, order)
        }
    }

    private fun compareKeys(a: ByteArray, b: ByteArray, order: MapKeyOrder): Int {
        if (order == MapKeyOrder.LengthFirst7049 && a.size != b.size) return a.size - b.size
        val n = minOf(a.size, b.size)
        for (i in 0 until n) {
            val c = (a[i].toInt() and 0xFF) - (b[i].toInt() and 0xFF)
            if (c != 0) return c
        }
        return a.size - b.size
    }
}

File: swift/Sources/CborCose/CborEncoder.swift
/// Map key ordering profile.
///
/// `bytewise8949` is the RFC 8949 §4.2.1 core deterministic order (default).
/// `lengthFirst7049` is the legacy RFC 7049 §3.9 canonical order (length, then bytewise) —
/// kept selectable because ISO 18013-5 artifacts in the wild differ; interop to be pinned in M4.
public enum MapKeyOrder: Sendable {
    case bytewise8949
    case lengthFirst7049
}

public enum CborEncodeError: Error, Equatable {
    case duplicateMapKey
    case invalidSimpleValue(UInt8)
}

public enum CborEncoder {

    public static func encode(_ value: Cbor, mapOrder: MapKeyOrder = .bytewise8949) throws -> [UInt8] {
        var out: [UInt8] = []
        try write(value, into: &out, order: mapOrder)
        return out
    }

    private static func head(_ major: UInt8, _ arg: UInt64, into out: inout [UInt8]) {
        let mt = major << 5
        switch arg {
        case ..<24:
            out.append(mt | UInt8(arg))
        case ...0xFF:
            out.append(mt | 24); out.append(UInt8(arg))
        case ...0xFFFF:
            out.append(mt | 25); out.append(UInt8(arg >> 8)); out.append(UInt8(arg & 0xFF))
        case ...0xFFFF_FFFF:
            out.append(mt | 26)
            var s = 24
            while s >= 0 { out.append(UInt8((arg >> UInt64(s)) & 0xFF)); s -= 8 }
        default:
            out.append(mt | 27)
            var s = 56
            while s >= 0 { out.append(UInt8((arg >> UInt64(s)) & 0xFF)); s -= 8 }
        }
    }

    private static func write(_ v: Cbor, into out: inout [UInt8], order: MapKeyOrder) throws {
        switch v {
        case let .uint(u):
            head(0, u, into: &out)
        case let .nint(n):
            head(1, n, into: &out)
        case let .bytes(b):
            head(2, UInt64(b.count), into: &out); out.append(contentsOf: b)
        case let .text(s):
            let b = Array(s.utf8)
            head(3, UInt64(b.count), into: &out); out.append(contentsOf: b)
        case let .array(items):
            head(4, UInt64(items.count), into: &out)
            for i in items { try write(i, into: &out, order: order) }
        case let .map(entries):
            try writeMap(entries, into: &out, order: order)
        case let .tagged(t, inner):
            head(6, t, into: &out); try write(inner, into: &out, order: order)
        case let .simple(s):
            try writeSimple(s, into: &out)
        case let .bool(b):
            out.append(b ? 0xF5 : 0xF4)
        case .null:
            out.append(0xF6)
        case .undefined:
            out.append(0xF7)
        case let .float(bits):
            writeFloat(bits: bits, into: &out)
        }
    }

    private static func writeSimple(_ s: UInt8, into out: inout [UInt8]) throws {
        if (20...23).contains(s) { throw CborEncodeError.invalidSimpleValue(s) }
        if s < 20 { out.append(0xE0 | s) } else { out.append(0xF8); out.append(s) }
    }

    private static func writeFloat(bits: UInt64, into out: inout [UInt8]) {
        let d = Double(bitPattern: bits)
        if d.isNaN { // canonical NaN
            out.append(contentsOf: [0xF9, 0x7E, 0x00])
            return
        }
        if let h = Half.exactBits(of: d) {
            out.append(0xF9); out.append(UInt8(h >> 8)); out.append(UInt8(h & 0xFF))
            return
        }
        let f = Float(d)
        if Double(f).bitPattern == bits {
            out.append(0xFA)
            let fb = f.bitPattern
            var s = 24
            while s >= 0 { out.append(UInt8((fb >> UInt32(s)) & 0xFF)); s -= 8 }
        } else {
            out.append(0xFB)
            var s = 56
            while s >= 0 { out.append(UInt8((bits >> UInt64(s)) & 0xFF)); s -= 8 }
        }
    }

    private static func writeMap(_ entries: [(Cbor, Cbor)], into out: inout [UInt8], order: MapKeyOrder) throws {
        head(5, UInt64(entries.count), into: &out)
        var encoded: [([UInt8], Cbor)] = []
        encoded.reserveCapacity(entries.count)
        for (k, v) in entries { encoded.append((try encode(k, mapOrder: order), v)) }
        let sorted = encoded.sorted { compareKeys($0.0, $1.0, order) < 0 }
        if sorted.count > 1 {
            for i in 1..<sorted.count where sorted[i - 1].0 == sorted[i].0 {
                throw CborEncodeError.duplicateMapKey
            }
        }
        for (kb, v) in sorted {
            out.append(contentsOf: kb)
            try write(v, into: &out, order: order)
        }
    }

    private static func compareKeys(_ a: [UInt8], _ b: [UInt8], _ order: MapKeyOrder) -> Int {
        if order == .lengthFirst7049 && a.count != b.count { return a.count - b.count }
        for (x, y) in zip(a, b) where x != y { return x < y ? -1 : 1 }
        return a.count - b.count
    }
}
